node/network/statement-distribution/src/lib.rs

```diff
@@ -0,0 +1,1376 @@
+		assert_eq!(knowledge.received_statements.len(), 3); // number of prior `Ok`s.
+
+		// Now make sure that the seconding limit is respected.
+		let hash_b: Hash = [2; 32].into();
+		let hash_c: Hash = [3; 32].into();
+
+		assert_eq!(
+			knowledge.receive(&(CompactStatement::Candidate(hash_b), 0), 3),
+			Ok(true),
+		);
+
+		assert_eq!(
+			knowledge.receive(&(CompactStatement::Candidate(hash_c), 0), 3),
+			Err(COST_UNEXPECTED_STATEMENT),
+		);
+
+		// Last, make sure that already-known statements are disregarded.
+		assert_eq!(
+			knowledge.receive(&(CompactStatement::Valid(hash_a), 2), 3),
+			Err(COST_DUPLICATE_STATEMENT),
+		);
+
+		assert_eq!(
+			knowledge.receive(&(CompactStatement::Candidate(hash_b), 0), 3),
+			Err(COST_DUPLICATE_STATEMENT),
+		);
+	}
+
+	#[test]
+	fn peer_view_update_sends_messages() {
+		let hash_a = [1; 32].into();
+		let hash_b = [2; 32].into();
+		let hash_c = [3; 32].into();
+
+		let candidate = {
+			let mut c = AbridgedCandidateReceipt::default();
+			c.relay_parent = hash_c;
+			c.parachain_index = 1.into();
+			c
+		};
+		let candidate_hash = candidate.hash();
+
+		let old_view = View(vec![hash_a, hash_b]);
+		let new_view = View(vec![hash_b, hash_c]);
+
+		let mut active_heads = HashMap::new();
+		let validators = vec![
+			Sr25519Keyring::Alice.public().into(),
+			Sr25519Keyring::Bob.public().into(),
+			Sr25519Keyring::Charlie.public().into(),
+		];
+
+		let session_index = 1;
+		let signing_context = SigningContext {
+			parent_hash: hash_c,
+			session_index,
+		};
+
+		let new_head_data = {
+			let mut data = ActiveHeadData::new(validators, session_index);
+
+			let noted = data.note_statement(SignedFullStatement::sign(
+				Statement::Seconded(candidate.clone()),
+				&signing_context,
+				0,
+				&Sr25519Keyring::Alice.pair().into(),
+			));
+
+			assert_matches!(noted, NotedStatement::Fresh(_));
+
+			let noted = data.note_statement(SignedFullStatement::sign(
+				Statement::Valid(candidate_hash),
+				&signing_context,
+				1,
+				&Sr25519Keyring::Bob.pair().into(),
+			));
+
+			assert_matches!(noted, NotedStatement::Fresh(_));
+
+			let noted = data.note_statement(SignedFullStatement::sign(
+				Statement::Valid(candidate_hash),
+				&signing_context,
+				2,
+				&Sr25519Keyring::Charlie.pair().into(),
+			));
+
+			assert_matches!(noted, NotedStatement::Fresh(_));
+
+			data
+		};
+
+		active_heads.insert(hash_c, new_head_data);
+
+		let mut peer_data = PeerData {
+			view: old_view,
+			view_knowledge: {
+				let mut k = HashMap::new();
+
+				k.insert(hash_a, Default::default());
+				k.insert(hash_b, Default::default());
+
+				k
+			},
+		};
+
+		let pool = ThreadPool::new().unwrap();
+		let (mut ctx, mut handle) = subsystem_test::make_subsystem_context(pool);
+		let peer = PeerId::random();
+
+		executor::block_on(async move {
+			update_peer_view_and_send_unlocked(
+				peer.clone(),
+				&mut peer_data,
+				&mut ctx,
+				&active_heads,
+				new_view.clone(),
+			).await.unwrap();
```

Comment:
```suggestion
			).await.expect("Must succeed for $reason; qed');
```

Reply:
There is an exception to our "panickers" rule for test code, so this should be OK.